reader_concurrency_semaphore: make CPU concurrency configurable

denesb opened the pull request:
The reader concurrency semaphore restricts the concurrency of reads that require CPU (intention: they read from the cache) to 1, meaning that if there is even a single active read which declares that it needs just CPU to proceed, no new read is admitted. This is meant to keep the concurrency of reads in the cache at 1. The idea is that concurrency in the cache is not useful: it just leads to the reactor rotating between these reads, all of the finishing later then they could if they were the only active read in the cache.
This was observed to backfire in the case where there reads from a single table are mostly very fast, but on some keys are very slow (hint: collection full of tombstones). In this case the slow read keeps up the fast reads in the queue, increasing the 99th percentile latencies significantly.

This series proposes to fix this, by making the CPU concurrency configurable. We don't like tunables like this and this is not a proper fix, but a workaround. The proper fix would be to allow to cut any page early, but we cannot cut a page in the middle of a row. We could maybe have a way of detecting slow reads and excluding them from the CPU concurrency. This would be a heuristic and it would be hard to get right. So in this series a robust and simple configurable is offered, which can be used on those few clusters which do suffer from the too strict concurrency limit. We have seen it in very few cases so far, so this doesn't seem to be wide-spread.

Fixes: https://github.com/scylladb/scylladb/issues/19017

This fixes a regression introduced in 5.0, so we have to backport to all currently supported releases

raphaelsc commented on reader_concurrency_semaphore.cc:
```diff
@@ -1265,7 +1274,7 @@ bool reader_concurrency_semaphore::has_available_units(const resources& r) const
 bool reader_concurrency_semaphore::all_need_cpu_permits_are_awaiting() const {
-    return _stats.need_cpu_permits == _stats.awaits_permits;
+    return (_stats.need_cpu_permits - _stats.awaits_permits) < _cpu_concurrency();
 }
```
now that concurrency can be higher than 1, should we rename all_need_cpu_permits_are_awaiting() to something like room_for_new_cpu_permit()?

for my education, even permits awaiting for I/O are still accounted in need_cpu_permits, right? so we won't have an underflow here.

nyh replied:
Right, the name of this function doesn't look apporpriate.

denesb replied:
Fixed in v2.

avikivity commented:
There's no reproducer attached. It's not easy to have reproducers of performance problems in test.py, but let's add something in test/manual.

nyh commented on db/config.cc:
```diff
@@ -1031,6 +1031,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , reader_concurrency_semaphore_kill_limit_multiplier(this, "reader_concurrency_semaphore_kill_limit_multiplier", liveness::LiveUpdate, value_status::Used, 4,
             "Start killing reads after their collective memory consumption goes above $normal_limit * $multiplier.")
+    , reader_concurrency_semaphore_cpu_concurrency(this, "reader_concurrency_semaphore_cpu_concurrency", liveness::LiveUpdate, value_status::Used, 1,
+            "Admit new reads while there are less than this number of request that need CPU.")
```
request**s**

denesb replied:
Fixed in v2.

nyh commented on test/boost/row_cache_test.cc:
```diff
@@ -4676,8 +4676,15 @@ SEASTAR_TEST_CASE(test_compact_range_tombstones_on_read) {
+            utils::updateable_value<uint32_t>(1),
+            utils::updateable_value<uint32_t>(1),
+            utils::updateable_value<uint32_t>(1),
+            reader_concurrency_semaphore::register_metrics::no);
```
I see a lot of cosmetic changes to the test, but I couldn't find an actual test for the new feature. Maybe it's difficult to test in a unit test? :-(

denesb replied:
Yes, I would need a latency test.

avikivity replied:
We can't test that in unit tests.

I suggest to add a new class of tests (or rather - expand the scope of our weekly perf-simple-query tests). These tests would run using an already-built package on a stable machine, all alone.

We can make them part of test/perf.

nyh reviewed (commented):
Beyond the general discussion about the concurrency of cpu-bound reads which I'm trying to hold in the issue, not in this PR, I have a question about this PR:
You suggested that the default of "1" isn't good which is why you are making it configurable. But you keep the default as the not-good 1. Why?  Are you suggesting that different users will need to set a different number depending on their workload? But,
1. This contradicts our self-adapting database mantra
2. How will users even know they need to set this configuration, and how will they choose a number?

I wonder if instead of a fixed number, we could have used some sort of function of the number of waiters, timeout, recent maximal latency, or something. But we can do all of this later. I'm not against adding this configuration for now just for emergencies (when a user needs this configuration, they will have it available).

avikivity commented:
> Beyond the general discussion about the concurrency of cpu-bound reads which I'm trying to hold in the issue, not in this PR, I have a question about this PR: You suggested that the default of "1" isn't good which is why you are making it configurable. But you keep the default as the not-good 1. Why? Are you suggesting that different users will need to set a different number depending on their workload? But,
> 
> 1. This contradicts our self-adapting database mantra
> 2. How will users even know they need to set this configuration, and how will they choose a number?
> 
> I wonder if instead of a fixed number, we could have used some sort of function of the number of waiters, timeout, recent maximal latency, or something. But we can do all of this later. I'm not against adding this configuration for now just for emergencies (when a user needs this configuration, they will have it available).

Having concurrency > 1 works if we have a wide distribution of compute time for pages. A small number of pages which take a long time to compute, and a large number of pages that take a small time to compute. So to adjust automatically, one needs to keep track of the time distribution.

In particular, it's important to avoid increasing concurrency when most of the requests take a long time, and when the CPU is overloaded. Perhaps it can be captured as "if usually we're not preempted, it's okay to increase concurrency occasionally, but if we're usually preempted, keep it low".

denesb commented:
> Beyond the general discussion about the concurrency of cpu-bound reads which I'm trying to hold in the issue, not in this PR, I have a question about this PR: You suggested that the default of "1" isn't good which is why you are making it configurable. But you keep the default as the not-good 1. Why?

We have introduced this change in 5.0 and in the 3.5 years that passed since, we just now seen the first case where this has proven to cause a regression in performance. There might be other cases we are not aware of, but it appears that the vast majority of workloads do not hit this regression. So I think we should stick to the previous behaviour (concurrency=1), and only change this new config if there is a problem with latencies and we strongly suspect this to be the cause.

That said maybe another number is just as good a default as 1, but I don't know how to determine this. So I will just leave it at the value which we know to be working well for the majority of cases.

> Are you suggesting that different users will need to set a different number depending on their workload? But,
> 
>     1. This contradicts our self-adapting database mantra
> 
>     2. How will users even know they need to set this configuration, and how will they choose a number?
> 
> 
> I wonder if instead of a fixed number, we could have used some sort of function of the number of waiters, timeout, recent maximal latency, or something. But we can do all of this later. I'm not against adding this configuration for now just for emergencies (when a user needs this configuration, they will have it available).

I don't know how to derive this number, not even statically, and much less at runtime.

Note that after our changes to cut pages after a certain number of tombstones, pretty much the only thing that can make reads in cache slow, is large collections, in particular collections with many tombstones. If we didn't have those, we wouldn't need this PR.

avikivity commented:
> 
> Note that after our changes to cut pages after a certain number of tombstones, pretty much the only thing that can make reads in cache slow, is large collections, in particular collections with many tombstones. If we didn't have those, we wouldn't need this PR.

Further to that, the patch helps only if only a small fraction of reads hit those large collections. If a larger fraction hits, increasing concurrency makes things worse, since the CPU has to thrash serving those requests concurrently.

denesb commented:
> Further to that, the patch helps only if only a small fraction of reads hit those large collections. If a larger fraction hits, increasing concurrency makes things worse, since the CPU has to thrash serving those requests concurrently.

Yes. Large collections are just plain bad.
An alternative solution would be if a read could detect reading a large collection of a certain size and would disqualify itself from the CPU based concurrency limit. It would still not help if such requests were common.

avikivity commented:
> > Further to that, the patch helps only if only a small fraction of reads hit those large collections. If a larger fraction hits, increasing concurrency makes things worse, since the CPU has to thrash serving those requests concurrently.
> 
> Yes. Large collections are just plain bad. An alternative solution would be if a read could detect reading a large collection of a certain size and would disqualify itself from the CPU based concurrency limit. It would still not help if such requests were common.

Let's start with the tunable, and only think about automation if we see > 0.3 tunings/year.

denesb commented:
> > Yes. Large collections are just plain bad. An alternative solution would be if a read could detect reading a large collection of a certain size and would disqualify itself from the CPU based concurrency limit. It would still not help if such requests were common.
> 
> Let's start with the tunable, and only think about automation if we see > 0.3 tunings/year.

Agree. I wasn't proposing to do the clever thing now, I was just thinking out loud.

denesb commented:
New in v2:
* Fix message in db::config.
* reader_concurrency_semaphore: s/all_need_cpu_permits_are_awaiting/cpu_concurrency_limit_reached/ (logic is now negated)

Still looking into how to write a test for this. Our current infra in perf/ is for throughput tests, and I need a latency test here.

raphaelsc commented:
> New in v2:
> 
> * Fix message in db::config.
> * reader_concurrency_semaphore: s/all_need_cpu_permits_are_awaiting/cpu_concurrency_limit_reached/ (logic is now negated)
> 
> Still looking into how to write a test for this. Our current infra in perf/ is for throughput tests, and I need a latency test here.

I think you can take a look at perf_cache_eviction for inspiration.

denesb commented:
> > New in v2:
> > 
> > * Fix message in db::config.
> > * reader_concurrency_semaphore: s/all_need_cpu_permits_are_awaiting/cpu_concurrency_limit_reached/ (logic is now negated)
> > 
> > Still looking into how to write a test for this. Our current infra in perf/ is for throughput tests, and I need a latency test here.
> 
> I think you can take a look at perf_cache_eviction for inspiration.

Thanks for the tip, I managed to come up with a test but it is working against this PR, increasing concurrency is making latencies worse.

mykaul commented:
@denesb - what's the next step here?

avikivity commented:
> > > New in v2:
> > > 
> > > * Fix message in db::config.
> > > * reader_concurrency_semaphore: s/all_need_cpu_permits_are_awaiting/cpu_concurrency_limit_reached/ (logic is now negated)
> > > 
> > > Still looking into how to write a test for this. Our current infra in perf/ is for throughput tests, and I need a latency test here.
> > 
> > 
> > I think you can take a look at perf_cache_eviction for inspiration.
> 
> Thanks for the tip, I managed to come up with a test but it is working against this PR, increasing concurrency is making latencies worse.

perf-simple-query with two workloads, one reading from a large-collection multi-row single-partition table with concurrency=1, the other reading from a key-value table with concurrency N, all from cache. With the default configuration throughput will be equal along all workers (and therefore latency will be high for the key/value table), with rcs cpu concurrency increased to N the key-value throughput will be higher by a factor of K and its latency will be lower by the same factor, where K is the number of times the large-collection workload is preempted per query.

denesb commented:
> @denesb - what's the next step here?

I'm trying to come up with the requested reproducer test and I'm failing. I have written the test, but no matter how much I tweaked the test, in every case, latencies go high or don't change, when I increase the semaphore concurrency, even though I implemented the workload that is known to benefit from the increase. I have already spent something like 3 days of time on this test and I'm not there yet.
There is a dtest PR (https://github.com/scylladb/scylla-dtest/pull/4477) with the original reproducer (written by @fee-mendes), just implemented in dtest. I will try to do some runs with that, to see if it can serve as a reproducer.

fruch commented:
> > @denesb - what's the next step here?
> 
> I'm trying to come up with the requested reproducer test and I'm failing. I have written the test, but no matter how much I tweaked the test, in every case, latencies go high or don't change, when I increase the semaphore concurrency, even though I implemented the workload that is known to benefit from the increase. I have already spent something like 3 days of time on this test and I'm not there yet.
> There is a dtest PR (https://github.com/scylladb/scylla-dtest/pull/4477) with the original reproducer (written by @fee-mendes), just implemented in dtest. I will try to do some runs with that, to see if it can serve as a reproducer.

I've refine @fee-mendes work especially for that purpose, but I've missed the thread here to mention it.

@denesb give me a ping if there is anything  I can help with.

avikivity commented:
> > @denesb - what's the next step here?
> 
> I'm trying to come up with the requested reproducer test and I'm failing. I have written the test, but no matter how much I tweaked the test, in every case, latencies go high or don't change, when I increase the semaphore concurrency, even though I implemented the workload that is known to benefit from the increase. I have already spent something like 3 days of time on this test and I'm not there yet. There is a dtest PR ([scylladb/scylla-dtest#4477](https://github.com/scylladb/scylla-dtest/pull/4477)) with the original reproducer (written by @fee-mendes), just implemented in dtest. I will try to do some runs with that, to see if it can serve as a reproducer.


The bright side is that it means the default configuration is good.

avikivity commented:
I believe that to get the increased concurrency to show benefit you have to tune the "slow" workload to take ~10ms for a page, but without stalling (say reading 100 rows where each row takes 0.1ms, or 1000 rows * 10us). Then you have 20 preemption opportunities per slow-workload page, which are ignored with the default configuration and taken advantage of with the non-default one.

denesb commented:
v3:
* Rebased and resolved conflicts (there are less updated constructor callers now, there is a simplified constructor which is backwards compatible).
* Added unit test for the new functionality.

Still no reproducer for the performance improvement. I had no luck with either the test/perf-based one, or the dtest one. Both show latency degradation with increased concurrency. It is possible that these canned tests, all running in the same process and/or in the same machine just don't reproduce the scenario we've seen in the field. In any case I argue that since this PR just introduces a tunable so we can react to bad situation in the field, and doesn't change the default, we don't really need a reproducer test (although It would certainly be nice).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10038/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10038/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/reader_concurrency_semaphore_test
#### Failed Tests (1/27):
- [test_reader_concurrency_semaphore_live_update_count](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10038/testReport/junit/boost/reader_concurrency_semaphore_test/test_reader_concurrency_semaphore_live_update_count) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_reader_concurrency_semaphore_live_update_count)
#### Build Details:
- Duration: 56 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10048/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10048/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/reader_concurrency_semaphore_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10048/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2330/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10350/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10048/)
#### Build Details:
- Duration: 3 hr 22 min
- Builder: spider4.cloudius-systems.com

nyh reviewed (approved):
I'm not happy about merging an "improvement" which doesn't show any improvement in any test case - and actually makes them worse - but since it just makes the hard-coded "1" configurable, and since there appears to be multiple emphatic requests from the field to make this "1" configurable, I think we can merge this - the downsides (a tiny bit added complexity) are fairly minor.

denesb commented:
I don't claim this PR to be an improvement, the reason I made this a tunable instead of just bumping it to some other value, is exactly because I'm not confident that the new value would be better. So I just want to give us the opportunity to intervene when we think the current value (1) doesn't work well.

bhalevy commented:
Note that we see `semaphore could admit new reads yet there are waiters` messages with this patch applied in the field.
Quoting @michoecho:
> The new errors are a result of a bug in the private patch. The patch broke an assumption held by a part of the code, and that part is complaining. I think this is a benign bug, but I am not sure about that. So we could want to hold off further upgrades until further review.

michoecho commented:
> Note that we see `semaphore could admit new reads yet there are waiters` messages with this patch applied in the field. Quoting @michoecho:
> 
> > The new errors are a result of a bug in the private patch. The patch broke an assumption held by a part of the code, and that part is complaining. I think this is a benign bug, but I am not sure about that. So we could want to hold off further upgrades until further review.

Scenario:
1. Two reads arrive together to the semaphore. The first one is placed in `_ready_list`. The second one is placed in `_wait_list`, (with `reason::ready_list`).
2. `execution_loop` wakes up, takes the first read out of `_ready_list` and starts it, and is preempted before it can admit the second read.
3. Third read arrives. It obtains `reason::all_ok` despite `_wait_list` being non-empty, so it screams about a suspected contradiction. (Even though there's nothing bad really happening. This assertion guards against a situation where we forget to admit waiters. This doesn't happen here — `execution_loop` is in the process of admitting waiters, it just got preempted).  

This wouldn't happen before the patch, because the third read in this situation would get queued with `reason::need_cpu_permits`.

bhalevy commented:
I think we need to call `maybe_admit_waiters` in 

> > Note that we see `semaphore could admit new reads yet there are waiters` messages with this patch applied in the field. Quoting @michoecho:
> > > The new errors are a result of a bug in the private patch. The patch broke an assumption held by a part of the code, and that part is complaining. I think this is a benign bug, but I am not sure about that. So we could want to hold off further upgrades until further review.
> 
> Scenario:
> 
> 1. Two reads arrive together to the semaphore. The first one is placed in `_ready_list`. The second one is placed in `_wait_list`, (with `reason::ready_list`).
> 2. `execution_loop` wakes up, takes the first read out of `_ready_list` and starts it, and is preempted before it can admit the second read.
> 3. Third read arrives. It obtains `reason::all_ok` despite `_wait_list` being non-empty, so it screams about a suspected contradiction. (Even though there's nothing bad really happening. This assertion guards against a situation where we forget to admit waiters. This doesn't happen here — `execution_loop` is in the process of admitting waiters, it just got preempted).
> 
> This wouldn't happen before the patch, because the third read in this situation would get queued with `reason::need_cpu_permits`.

I think that the following patch should address the above race:
```diff
diff --git a/reader_concurrency_semaphore.cc b/reader_concurrency_semaphore.cc
index afd6c88d0a..98bef9598c 100644
--- a/reader_concurrency_semaphore.cc
+++ b/reader_concurrency_semaphore.cc
@@ -925,6 +925,9 @@ future<> reader_concurrency_semaphore::execution_loop() noexcept {
         }
 
         while (!_ready_list.empty()) {
+            if (need_preempt()) {
+                co_await coroutine::maybe_yield();
+            }
             auto& permit = _ready_list.front();
             dequeue_permit(permit);
             permit.on_executing();
@@ -937,10 +940,6 @@ future<> reader_concurrency_semaphore::execution_loop() noexcept {
             } catch (...) {
                 e.pr.set_exception(std::current_exception());
             }
-
-            if (need_preempt()) {
-                co_await coroutine::maybe_yield();
-            }
         }
 
         maybe_admit_waiters();
```

vladzcloudius commented:
> > Note that we see `semaphore could admit new reads yet there are waiters` messages with this patch applied in the field. Quoting @michoecho:
> > > The new errors are a result of a bug in the private patch. The patch broke an assumption held by a part of the code, and that part is complaining. I think this is a benign bug, but I am not sure about that. So we could want to hold off further upgrades until further review.
> 
> Scenario:
> 
>     1. Two reads arrive together to the semaphore. The first one is placed in `_ready_list`. The second one is placed in `_wait_list`, (with `reason::ready_list`).
> 
>     2. `execution_loop` wakes up, takes the first read out of `_ready_list` and starts it, and is preempted before it can admit the second read.
> 
>     3. Third read arrives. It obtains `reason::all_ok` despite `_wait_list` being non-empty, so it screams about a suspected contradiction. (Even though there's nothing bad really happening. This assertion guards against a situation where we forget to admit waiters. This doesn't happen here — `execution_loop` is in the process of admitting waiters, it just got preempted).

Why the second reader in (1) is put in the _waiter_list and not in the _ready_list if there are enough permits to do so, @michoecho ?

> 
> 
> This wouldn't happen before the patch, because the third read in this situation would get queued with `reason::need_cpu_permits`.

denesb commented:
> Why the second reader in (1) is put in the _waiter_list and not in the _ready_list if there are enough permits to do so, @michoecho ?

The size of the _ready_list is limited to the CPU concurrency, which used to be hardcoded to 1. The patch didn't update this, hence the confusion above.

denesb commented:
@bhalevy the fix is this:
```
diff --git a/reader_concurrency_semaphore.cc b/reader_concurrency_semaphore.cc
index afd6c88d0a..70e7cea2f6 100644
--- a/reader_concurrency_semaphore.cc
+++ b/reader_concurrency_semaphore.cc
@@ -1354,7 +1354,7 @@ reader_concurrency_semaphore::can_admit_read(const reader_permit::impl& permit)
         return {can_admit::yes, reason::all_ok};
     }
 
-    if (!_ready_list.empty()) {
+    if (_ready_list.size() >= _cpu_concurrency()) {
         return {can_admit::no, reason::ready_list};
     }
```

I will send a PR.

bhalevy commented:
> @bhalevy the fix is this:
> 
> ```diff
> diff --git a/reader_concurrency_semaphore.cc b/reader_concurrency_semaphore.cc
> index afd6c88d0a..70e7cea2f6 100644
> --- a/reader_concurrency_semaphore.cc
> +++ b/reader_concurrency_semaphore.cc
> @@ -1354,7 +1354,7 @@ reader_concurrency_semaphore::can_admit_read(const reader_permit::impl& permit)
>          return {can_admit::yes, reason::all_ok};
>      }
>  
> -    if (!_ready_list.empty()) {
> +    if (_ready_list.size() >= _cpu_concurrency()) {
>          return {can_admit::no, reason::ready_list};
>      }
> ```
> 
> I will send a PR.

Makes sense, but still consider the yielding order change as well, Otherwise we might end up with _ready_list.empty() &&  !_wait_list.empty() if the execution loop yields just at that time.
It can be officially handled and expected too, and the message about it be deleted, but let's not leave both in place.

bhalevy commented:
In fact, even with your patch, if the execution loop yields when the ready list size is less than cpu concurrency, we can end up with can_admit::yes and _wait_list not empty?

Given that, this state should be officially supported, and the diagnostics message dropped

denesb commented:
I now remember why `_ready_list` had a hardcoded size of 1. This is not because CPU concurrency is 1, but because otherwise the CPU concurrency can be completely bypassed. This is still the case. I need to reconsider how to handle this.

tchaikov commented on reader_concurrency_semaphore.cc:
```diff
@@ -1274,8 +1274,8 @@ bool reader_concurrency_semaphore::has_available_units(const resources& r) const
-bool reader_concurrency_semaphore::all_need_cpu_permits_are_awaiting() const {
-    return _stats.need_cpu_permits == _stats.awaits_permits;
+bool reader_concurrency_semaphore::cpu_concurrency_limit_reached() const {
+    return (_stats.need_cpu_permits - _stats.awaits_permits) >= _cpu_concurrency();
```
this might run into unsigned integer underflow, see https://godbolt.org/z/TaxMYxMjz

so probably we should put something like:

```c++
uint64_t provisioned = _stats.awaits_permits + _cpu_concurrency();
return _stats.need_cpu_permits >= provisioned;
```

denesb replied:
This suggests that there is a bug somewhere: all permits that are in awaits state, are also accounted in the `need_cpu_permits` state too.
See https://github.com/scylladb/scylladb/blob/591876b44e5a38052c9ab2d9d4f219bcf87275a9/reader_concurrency_semaphore.cc#L1543-L1547

michoecho replied:
No, because `await` is a sub-state of `need_cpu`, `_stats.need_cpu_permits >= _stats.awaits_permits` is an invariant, protected by asserts.

michoecho replied:
My comment raced with Botond's, it was directed at @tchaikov.

tchaikov replied:
Thanks guys. I see, so it's a red herring. I am chasing a bug in this area. And thus caught my attention.
